### ISSUE TITLE
F-Droid version is outdated

### ISSUE THREAD

**Issue by CrimsonFork** [hrydgard/ppsspp#9112]
It's still 1.1.1.0

**Comment by hrydgard** [comment]
The F-Droid build is not maintained by the project, it was done by some volunteer. It would be great if a new one would step up.

**Comment by OffifialMITX** [comment]
Well, I don't know how it works, but is it really so complicated?

**Comment by hrydgard** [comment]
It's probably not complicated, I just don't really know what exactly F-droid is or how it works, other than it being some sort of repository for open source Android apps. Hasn't really been on my radar.

**Comment by Krakn** [comment]
I'm looking into this.

**Comment by OffifialMITX** [comment]
Finally! Someone! Our fate rests upon you!

**Comment by Krakn** [comment]
@hrydgard 

If you have a few minutes, I've been doing some digging...

https://f-droid.org/wiki/page/Inclusion_Policy

> * We cannot build apps using Google's proprietary "play-services". Please talk to upstream about an untainted build flavor (either using microg or removing non-free deps completely).
> * We cannot build apps using proprietary tracking/analytic dependencies like Crashlytics and Firebase. Please talk to upstream about an untainted build flavor (either using a FLOSS analytics software like ACRA or by removing non-free deps completely).
> * We cannot build apps using proprietary ad libraries. We have nothing against advertisements (see AntiFeatures), but they must be provided by a FLOSS compatible way.
> * We cannot build apps requiring non-free buildtools, including Oracle's JDK or some pre-release toolchains.

If I remember correctly, PPSSPP used (might still be using) analytics libraries? I'm not sure I can answer the rest of those without digging into the code a bit more...

That said, moving the build system over from those <code>ab.cmd</code> and <code>ab.sh</code> scripts would be quite useful... Moving to something like [Android Studio](https://developer.android.com/studio/index.html) would allow simplified building from the command line with tools like [Gradle](https://gradle.org/).

Links:

* https://developer.android.com/studio/build/building-cmdline.html

* https://developer.android.com/studio/intro/migrate.html

**Comment by Orphis** [comment]
We have an experimental branch to build with Gradle + CMake actually. It's waiting on some packages with bug fixes we need being promoted from beta to release. Then, you will be able to build using Gradle.
But the aforementioned scripts are simple enough, even if they aren't standard. What issues do you have with them?

**Comment by Krakn** [comment]
No problems, just increased simplicity.

More "standardized".

**Comment by hrydgard** [comment]
PPSSPP no longer use anything from Google Play services. No analytics. So we should be fine there.

**Comment by Krakn** [comment]
"Blocker" - https://github.com/hrydgard/ppsspp/pull/9038 has been merged.

I'll start work on this shortly.

**Comment by hrydgard** [comment]
Hm, so what APK signing is used on f-droid? Debug signing or what? If you need to add an additional "flavor" to the gradle file with some special signature that's alright.

**Comment by Krakn** [comment]
@hrydgard would you be opposed to me migrating the build system over to [gradle](https://gradle.org/)?

Additional deatils (as they pertain to ppsspp) here: https://developer.android.com/studio/intro/migrate.html

**Comment by hrydgard** [comment]
@Krakn The current build system for Android is already gradle+cmake (though we do keep the old ant/Eclipse build as well). Or do you mean just switching the f-droid build to use gradle? Go right ahead. I'm not sure how f-droid does signing, btw?

**Comment by Krakn** [comment]
I was using my outdated fork that was missing the cmake + gradle commits... My bad on that one.

I'm looking into how F-Droid handles signing now.

**Comment by hrydgard** [comment]
There's support in the gradle scripts for using a certificate specified in .gradleproperties, if f-droid uses its own cert. If debug keys is enough, just do gradle assembleOptimized.

**Comment by Krakn** [comment]
### Status:

- [x] Building on local machine
- [x] F-Droid build environment set up
- [ ] Building with F-Droid's build servers
- [ ] Published to F-Droid's repositories

**Comment by Lionirdeadman** [comment]
@devopsdeluxe Any news?

**Comment by devopsdeluxe** [comment]
I had some large life changes recently and this fell to the wayside. :(

I'll see if I can get things completed this weekend!

**Comment by Lionirdeadman** [comment]
Sorry to hear..

Well, I hope you get things going! Thanks.

**Comment by devopsdeluxe** [comment]
Update: https://gitlab.com/fdroid/fdroiddata/issues/1030

**Comment by devopsdeluxe** [comment]
I just need to test the built apk and we should be in business. I'll ping hrydgard if I need singing stuff.

**Comment by devopsdeluxe** [comment]
I've picked this up again and am nearing completion: https://gitlab.com/fdroid/fdroiddata/issues/1030

I'm hoping to have the build published by the end of this week.

**Comment by unknownbrackets** [comment]
Just a warning: we're planning to release v1.6.0 sometime in the near future.  See #10773.

-[Unknown]

**Comment by devopsdeluxe** [comment]
@unknownbrackets Thanks for the head up!

_Assuming_ I created a sane metadata entry on fdroid's side, the version bump (1.5.4 -> 1.6.0) should just be a quick copy & paste job.

**Comment by devopsdeluxe** [comment]
https://gitlab.com/fdroid/fdroiddata/merge_requests/3083

**Comment by Emulatorer** [comment]
UP

**Comment by OffifialMITX** [comment]
Yay

On July 10, 2018 12:18:56 PM GMT+02:00, Emulatorer <notifications@github.com> wrote:
>UP
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/hrydgard/ppsspp/issues/9112#issuecomment-403773921

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by Florin9doi** [comment]
Stubs for com/bda/controller ?

https://github.com/Florin9doi/ppsspp/commit/b0ec49309eadc1218652b7be5a513e2a7cd2494f ?

**Comment by unknownbrackets** [comment]
That seems like an option imho (assuming it works.)  Though not sure about "Maga".  Maybe "MogaStubs" would be better, or something.

-[Unknown]

**Comment by Florin9doi** [comment]
https://github.com/hrydgard/ppsspp/pull/11545 has been merged but F-droid version should  probably wait until a new tag is added

**Comment by Florin9doi** [comment]
@devopsdeluxe, can you update your PR to use the latest PPSSPP tag and to execute `build=gradle assembleRelease -Pf_droid`?

**Comment by devopsdeluxe** [comment]
When someone has a minute, I'm getting the following error while building v1.8.0:

```
22:21:26 drb@aether:/home/drb/Git/ppsspp/android
 + gradle assembleRelease -Pf_droid

> Configure project :android
Overriding Android Version Name, Code: 1.8.0 18000000
/home/drb/Git/ppsspp/assets

> Task :android:preBuild UP-TO-DATE
> Task :android:preGoldReleaseBuild UP-TO-DATE
> Task :android:compileGoldReleaseAidl NO-SOURCE
> Task :android:compileGoldReleaseRenderscript UP-TO-DATE
> Task :android:checkGoldReleaseManifest UP-TO-DATE
> Task :android:generateGoldReleaseBuildConfig UP-TO-DATE
> Task :android:prepareLintJar UP-TO-DATE
> Task :android:generateGoldReleaseSources UP-TO-DATE
> Task :android:javaPreCompileGoldRelease UP-TO-DATE
> Task :android:mainApkListPersistenceGoldRelease UP-TO-DATE
> Task :android:generateGoldReleaseResValues UP-TO-DATE
> Task :android:generateGoldReleaseResources UP-TO-DATE
> Task :android:mergeGoldReleaseResources UP-TO-DATE
> Task :android:createGoldReleaseCompatibleScreenManifests UP-TO-DATE
> Task :android:processGoldReleaseManifest UP-TO-DATE
> Task :android:processGoldReleaseResources UP-TO-DATE
> Task :android:compileGoldReleaseJavaWithJavac UP-TO-DATE
> Task :android:generateJsonModelGoldRelease UP-TO-DATE

> Task :android:externalNativeBuildGoldRelease
Build armips-bin x86_64
ninja: no work to do.
Build armipstests x86_64
ninja: no work to do.
Build ppsspp_jni x86_64
[1/1] Generating something_that_never_exists
-- Found Git: /usr/bin/git (found version "2.21.0") 
Build armipstests x86
ninja: no work to do.
Build armips-bin x86
ninja: no work to do.
Build ppsspp_jni x86
[1/1] Generating something_that_never_exists
-- Found Git: /usr/bin/git (found version "2.21.0") 
Build armipstests arm64-v8a
ninja: no work to do.
Build ppsspp_jni arm64-v8a
[1/1] Generating something_that_never_exists
-- Found Git: /usr/bin/git (found version "2.21.0") 
Build armips-bin arm64-v8a
ninja: no work to do.
Build armipstests armeabi-v7a
ninja: no work to do.
Build ppsspp_jni armeabi-v7a
[1/329] Building CXX object CMakeFiles/native.dir/ext/native/ext/jpge/jpge.cpp.o
[2/329] Generating something_that_never_exists
[3/329] Building CXX object ext/glslang/hlsl/CMakeFiles/HLSL.dir/hlslParseHelper.cpp.o
[4/329] Linking CXX static library ext/glslang/SPIRV/libSPVRemapper.a
[5/329] Linking CXX static library ext/SPIRV-Cross-build/libspirv-cross-core.a
[6/329] Building CXX object ext/discord-rpc-build/CMakeFiles/discord-rpc.dir/__/discord-rpc/src/discord_register_linux.cpp.o
-- Found Git: /usr/bin/git (found version "2.21.0") 
[7/329] Building CXX object ext/SPIRV-Cross-build/CMakeFiles/spirv-cross-glsl.dir/__/SPIRV-Cross/spirv_glsl.cpp.o
[8/329] Linking CXX static library ext/discord-rpc-build/libdiscord-rpc.a
[9/329] Linking CXX static library libnative.a
FAILED: /opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++  --target=armv7-none-linux-androideabi16 --gcc-toolchain=/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64 --sysroot=/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot  -DAMD_EXTENSIONS -DENABLE_HLSL -DENABLE_OPT=0 -DGLSLANG_OSINCLUDE_UNIX -DGOLD -DMOBILE_DEVICE -DNV_EXTENSIONS -DUSING_GLES2 -D__STDC_CONSTANT_MACROS -I../../../../../ext/native -I../../../../../ -g -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -fno-addrsig -march=armv7-a -mthumb -mfpu=neon -Wa,--noexecstack -Wformat -Werror=format-security   -std=gnu++11 -Wno-switch -Wno-uninitialized -Oz -DNDEBUG  -O3 -D_NDEBUG -fPIC   -Wno-multichar -Wno-deprecated-register -fsigned-char -mfpu=neon -include /home/drb/Git/ppsspp/ppsspp_config.h -fno-strict-aliasing -Wall -Wuninitialized -Wunused -Wunused-local-typedefs -Wunused-parameter -Wunused-value -Wunused-variable -Wno-reorder -std=c++11 -MD -MT ext/glslang/hlsl/CMakeFiles/HLSL.dir/hlslParseHelper.cpp.o -MF ext/glslang/hlsl/CMakeFiles/HLSL.dir/hlslParseHelper.cpp.o.d -o ext/glslang/hlsl/CMakeFiles/HLSL.dir/hlslParseHelper.cpp.o -c /home/drb/Git/ppsspp/ext/glslang/hlsl/hlslParseHelper.cpp
Stack dump:
0.	Program arguments: /opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++ -cc1 -triple thumbv7-none-linux-android16 -emit-obj -mnoexecstack -disable-free -disable-llvm-verifier -discard-value-names -main-file-name hlslParseHelper.cpp -mrelocation-model pic -pic-level 2 -mthread-model posix -mdisable-fp-elim -relaxed-aliasing -masm-verbose -mconstructor-aliases -munwind-tables -fuse-init-array -target-cpu generic -target-feature +soft-float-abi -target-feature -fp-only-sp -target-feature -d16 -target-feature +vfp3 -target-feature -fp16 -target-feature -vfp4 -target-feature -fp-armv8 -target-feature +neon -target-feature -crypto -target-abi aapcs-linux -mfloat-abi soft -fallow-half-arguments-and-returns -dwarf-column-info -debug-info-kind=limited -dwarf-version=4 -debugger-tuning=gdb -ffunction-sections -fdata-sections -coverage-notes-file /home/drb/Git/ppsspp/android/.externalNativeBuild/cmake/goldRelease/armeabi-v7a/ext/glslang/hlsl/CMakeFiles/HLSL.dir/hlslParseHelper.cpp.gcno -resource-dir /opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/lib64/clang/8.0.7 -dependency-file ext/glslang/hlsl/CMakeFiles/HLSL.dir/hlslParseHelper.cpp.o.d -sys-header-deps -MT ext/glslang/hlsl/CMakeFiles/HLSL.dir/hlslParseHelper.cpp.o -include /home/drb/Git/ppsspp/ppsspp_config.h -D AMD_EXTENSIONS -D ENABLE_HLSL -D ENABLE_OPT=0 -D GLSLANG_OSINCLUDE_UNIX -D GOLD -D MOBILE_DEVICE -D NV_EXTENSIONS -D USING_GLES2 -D __STDC_CONSTANT_MACROS -I ../../../../../ext/native -I ../../../../../ -D ANDROID -D NDEBUG -D _NDEBUG -isysroot /opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot -internal-isystem /opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/c++/v1 -internal-isystem /opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/local/include -internal-isystem /opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/lib64/clang/8.0.7/include -internal-externc-isystem /opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/arm-linux-androideabi -inter
nal-externc-isystem /opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot/include -internal-externc-isystem /opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include -O3 -Wformat -Werror=format-security -Wno-switch -Wno-uninitialized -Wno-multichar -Wno-deprecated-register -Wall -Wuninitialized -Wunused -Wunused-local-typedefs -Wunused-parameter -Wunused-value -Wunused-variable -Wno-reorder -std=c++11 -fdeprecated-macro -fdebug-compilation-dir /home/drb/Git/ppsspp/android/.externalNativeBuild/cmake/goldRelease/armeabi-v7a -ferror-limit 19 -fmessage-length 0 -stack-protector 2 -fobjc-runtime=gcc -fcxx-exceptions -fexceptions -fdiagnostics-show-option -vectorize-loops -vectorize-slp -o ext/glslang/hlsl/CMakeFiles/HLSL.dir/hlslParseHelper.cpp.o -x c++ /home/drb/Git/ppsspp/ext/glslang/hlsl/hlslParseHelper.cpp 
1.	<eof> parser at end of file
2.	Code generation
3.	Running pass 'Function Pass Manager' on module '/home/drb/Git/ppsspp/ext/glslang/hlsl/hlslParseHelper.cpp'.
4.	Running pass 'Simple Register Coalescing' on function '@_ZN7glslang16HlslParseContext17remapEntryPointIOERNS_9TFunctionERPNS_9TVariableERNS_7TVectorIS4_EES8_'
#0 0x000000000242f066 llvm::sys::PrintStackTrace(llvm::raw_ostream&) (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x242f066)
#1 0x000000000242dca0 llvm::sys::RunSignalHandlers() (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x242dca0)
#2 0x000000000242f34f (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x242f34f)
#3 0x00007fd515a614d0 __restore_rt (/usr/lib/libpthread.so.0+0x124d0)
#4 0x0000000001e43747 llvm::LiveRange::join(llvm::LiveRange&, int const*, int const*, llvm::SmallVectorImpl<llvm::VNInfo*>&) (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x1e43747)
#5 0x0000000001eec54f (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x1eec54f)
#6 0x0000000001ee8d18 (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x1ee8d18)
#7 0x0000000001ee8686 (/
opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x1ee8686)
#8 0x0000000001ee83cb (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x1ee83cb)
#9 0x0000000001ee7d85 (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x1ee7d85)
#10 0x0000000001ee7a57 (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x1ee7a57)
#11 0x0000000001e79ec3 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x1e79ec3)
#12 0x000000000209208b llvm::FPPassManager::runOnFunction(llvm::Function&) (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x209208b)
#13 0x00000000020924e3 llvm::FPPassManager::runOnModule(llvm::Module&) (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x20924e3)
#14 0x00000000020928ad (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x20928ad)
#15 0x000000000209265d llvm::legacy::PassManagerImpl::run(llvm::Module&) (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x209265d)
#16 0x0000000002755c1c (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x2755c1c)
#17 0x0000000002753184 clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::DataLayout const&, llvm::Module*, clang::BackendAction, std::__1::unique_ptr<llvm::raw_pwrite_stream, std::__1::default_delete<llvm::raw_pwrite_stream> >) (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x2753184)
#18 0x0000000002936591 (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x2936591)
#19 0x000000000307b7a8 clang::ParseAST(clang::Sema&, bool, bool) (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x307b7a8)
#20 0x00000000028d0cf5 clang::FrontendAction::Execute() (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x28d0cf5)
#21 0x000000000286db67 clang::
CompilerInstance::ExecuteAction(clang::FrontendAction&) (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x286db67)
#22 0x0000000002932ff5 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x2932ff5)
#23 0x00000000018876d4 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x18876d4)
#24 0x0000000001883f22 main (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x1883f22)
#25 0x00007fd51544fce3 __libc_start_main (/usr/lib/libc.so.6+0x23ce3)
#26 0x0000000001883029 _start (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x1883029)
clang++: error: unable to execute command: Segmentation fault (core dumped)
clang++: error: clang frontend command failed due to signal (use -v to see invocation)
Android (5220042 based on r346389c) clang version 8.0.7 (https://android.googlesource.com/toolchain/clang b55f2d4ebfd35bf643d27dbca1bb228957008617) (https://android.googlesource.com/toolchain/llvm 3c393fe7a7e13b0fba4ac75a01aa683d7a5b11cd) (based on LLVM 8.0.7svn)
Target: armv7-none-linux-android16
Thread model: posix
InstalledDir: /opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin
clang++: note: diagnostic msg: PLEASE submit a bug report to https://github.com/android-ndk/ndk/issues and include the crash backtrace, preprocessed source, and associated run script.
clang++: note: diagnostic msg: 
********************
PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang++: note: diagnostic msg: /tmp/hlslParseHelper-3e494b.cpp
clang++: note: diagnostic msg: /tmp/hlslParseHelper-3e494b.sh
clang++: note: diagnostic msg: 
********************
ninja: build stopped: subcommand failed.

> Task :android:externalNativeBuildGoldRelease FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':android:externalNativeBuildGoldRelease'.
> Build command failed.
  Error while executing process /opt/android-sdk/cmake/3.6.4111459/bin/cmake with arguments {--build /home/drb/Git/ppsspp/android/.externalNativeBuild/cmake/goldRelease/armeabi-v7a --target ppsspp_jni}
  [1/329] Building CXX object CMakeFiles/native.dir/ext/native/ext/jpge/jpge.cpp.o
  [2/329] Generating something_that_never_exists
  [3/329] Building CXX object ext/glslang/hlsl/CMakeFiles/HLSL.dir/hlslParseHelper.cpp.o
  [4/329] Linking CXX static library ext/glslang/SPIRV/libSPVRemapper.a
  [5/329] Linking CXX static library ext/SPIRV-Cross-build/libspirv-cross-core.a
  [6/329] Building CXX object ext/discord-rpc-build/CMakeFiles/discord-rpc.dir/__/discord-rpc/src/discord_register_linux.cpp.o
  -- Found Git: /usr/bin/git (found version "2.21.0") 
  [7/329] Building CXX object ext/SPIRV-Cross-build/CMakeFiles/spirv-cross-glsl.dir/__/SPIRV-Cross/spirv_glsl.cpp.o
  [8/329] Linking CXX static library ext/discord-rpc-build/libdiscord-rpc.a
  [9/329] Linking CXX static library libnative.a
  FAILED: /opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++  --target=armv7-none-linux-androideabi16 --gcc-toolchain=/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64 --sysroot=/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot  -DAMD_EXTENSIONS -DENABLE_HLSL -DENABLE_OPT=0 -DGLSLANG_OSINCLUDE_UNIX -DGOLD -DMOBILE_DEVICE -DNV_EXTENSIONS -DUSING_GLES2 -D__STDC_CONSTANT_MACROS -I../../../../../ext/native -I../../../../../ -g -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -fno-addrsig -march=armv7-a -mthumb -mfpu=neon -Wa,--noexecstack -Wformat -Werror=format-security   -std=gnu++11 -Wno-switch -Wno-uninitialized -Oz -DNDEBUG  -O3 -D_NDEBUG -fPIC   -Wno-multichar -Wno-deprecated-register -fsigned-char -mfpu=neon -include /home/drb/Git/ppsspp/ppsspp_config.h -fno-strict-aliasing -Wall -Wuninitialized -Wunused -Wunused-local-typedefs -Wunused-parameter -Wunused-value -Wunused-variable -Wno-reorder -std=c++11 -MD -MT ext/glslang/hlsl/CMakeFiles/HLSL.dir/hlslParseHelper.cpp.o -MF ext/glslang/hlsl/CMakeFiles/HLSL.dir/hlslParseHelper.cpp.o.d -o ext/glslang/hlsl/CMakeFiles/HLSL.dir/hlslParseHelper.cpp.o -c /home/drb/Git/ppsspp/ext/glslang/hlsl/hlslParseHelper.cpp
  Stack dump:
  0.	Program arguments: /opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++ -cc1 -triple thumbv7-none-linux-android16 -emit-obj -mnoexecstack -disable-free -disable-llvm-verifier -discard-value-names -main-file-name hlslParseHelper.cpp -mrelocation-model pic -pic-level 2 -mthread-model posix -mdisable-fp-elim -relaxed-aliasing -masm-verbose -mconstructor-aliases -munwind-tables -fuse-init-array -target-cpu generic -target-feature +soft-float-abi -target-feature -fp-only-sp -target-feature -d16 -target-feature +vfp3 -target-feature -fp16 -target-feature -vfp4 -target-feature -fp-armv8 -target-feature +neon -target-feature -crypto -target-abi aapcs-linux -mfloat-abi soft -fallow-half-arguments-and-returns -dwarf-column-info -debug-info-kind=limited -dwarf-version=4 -debugger-tuning=gdb -ffunction-sections -fdata-sections -coverage-notes-file /home/drb/Git/ppsspp/android/.externalNativeBuild/cmake/goldRelease/armeabi-v7a/ext/glslang/hlsl/CMakeFiles/HLSL.dir/hlslParseHelper.cpp.gcno -resource-dir /opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/lib64/clang/8.0.7 -dependency-file ext/glslang/hlsl/CMakeFiles/HLSL.dir/hlslParseHelper.cpp.o.d -sys-header-deps -MT ext/glslang/hlsl/CMakeFiles/HLSL.dir/hlslParseHelper.cpp.o -include /home/drb/Git/ppsspp/ppsspp_config.h -D AMD_EXTENSIONS -D ENABLE_HLSL -D ENABLE_OPT=0 -D GLSLANG_OSINCLUDE_UNIX -D GOLD -D MOBILE_DEVICE -D NV_EXTENSIONS -D USING_GLES2 -D __STDC_CONSTANT_MACROS -I ../../../../../ext/native -I ../../../../../ -D ANDROID -D NDEBUG -D _NDEBUG -isysroot /opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot -internal-isystem /opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/c++/v1 -internal-isystem /opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/local/include -internal-isystem /opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/lib64/clang/8.0.7/include -internal-externc-isystem /opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/arm-linux-androideabi -internal-externc-isystem /opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot/include -internal-externc-isystem /opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include -O3 -Wformat -Werror=format-security -Wno-switch -Wno-uninitialized -Wno-multichar -Wno-deprecated-register -Wall -Wuninitialized -Wunused -Wunused-local-typedefs -Wunused-parameter -Wunused-value -Wunused-variable -Wno-reorder -std=c++11 -fdeprecated-macro -fdebug-compilation-dir /home/drb/Git/ppsspp/android/.externalNativeBuild/cmake/goldRelease/armeabi-v7a -ferror-limit 19 -fmessage-length 0 -stack-protector 2 -fobjc-runtime=gcc -fcxx-exceptions -fexceptions -fdiagnostics-show-option -vectorize-loops -vectorize-slp -o ext/glslang/hlsl/CMakeFiles/HLSL.dir/hlslParseHelper.cpp.o -x c++ /home/drb/Git/ppsspp/ext/glslang/hlsl/hlslParseHelper.cpp 
  1.	<eof> parser at end of file
  2.	Code generation
  3.	Running pass 'Function Pass Manager' on module '/home/drb/Git/ppsspp/ext/glslang/hlsl/hlslParseHelper.cpp'.
  4.	Running pass 'Simple Register Coalescing' on function '@_ZN7glslang16HlslParseContext17remapEntryPointIOERNS_9TFunctionERPNS_9TVariableERNS_7TVectorIS4_EES8_'
  #0 0x000000000242f066 llvm::sys::PrintStackTrace(llvm::raw_ostream&) (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x242f066)
  #1 0x000000000242dca0 llvm::sys::RunSignalHandlers() (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x242dca0)
  #2 0x000000000242f34f (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x242f34f)
  #3 0x00007fd515a614d0 __restore_rt (/usr/lib/libpthread.so.0+0x124d0)
  #4 0x0000000001e43747 llvm::LiveRange::join(llvm::LiveRange&, int const*, int const*, llvm::SmallVectorImpl<llvm::VNInfo*>&) (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x1e43747)
  #5 0x0000000001eec54f (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x1eec54f)
  #6 0x0000000001ee8d18 (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x1ee8d18)
  #7 0x0000000001ee8686 (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x1ee8686)
  #8 0x0000000001ee83cb (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x1ee83cb)
  #9 0x0000000001ee7d85 (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x1ee7d85)
  #10 0x0000000001ee7a57 (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x1ee7a57)
  #11 0x0000000001e79ec3 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x1e79ec3)
  #12 0x000000000209208b llvm::FPPassManager::runOnFunction(llvm::Function&) (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x209208b)
  #13 0x00000000020924e3 llvm::FPPassManager::runOnModule(llvm::Module&) (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x20924e3)
  #14 0x00000000020928ad (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x20928ad)
  #15 0x000000000209265d llvm::legacy::PassManagerImpl::run(llvm::Module&) (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x209265d)
  #16 0x0000000002755c1c (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x2755c1c)
  #17 0x0000000002753184 clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::DataLayout const&, llvm::Module*, clang::BackendAction, std::__1::unique_ptr<llvm::raw_pwrite_stream, std::__1::default_delete<llvm::raw_pwrite_stream> >) (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x2753184)
  #18 0x0000000002936591 (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x2936591)
  #19 0x000000000307b7a8 clang::ParseAST(clang::Sema&, bool, bool) (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x307b7a8)
  #20 0x00000000028d0cf5 clang::FrontendAction::Execute() (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x28d0cf5)
  #21 0x000000000286db67 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x286db67)
  #22 0x0000000002932ff5 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x2932ff5)
  #23 0x00000000018876d4 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x18876d4)
  #24 0x0000000001883f22 main (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x1883f22)
  #25 0x00007fd51544fce3 __libc_start_main (/usr/lib/libc.so.6+0x23ce3)
  #26 0x0000000001883029 _start (/opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x1883029)
  clang++: error: unable to execute command: Segmentation fault (core dumped)
  clang++: error: clang frontend command failed due to signal (use -v to see invocation)
  Android (5220042 based on r346389c) clang version 8.0.7 (https://android.googlesource.com/toolchain/clang b55f2d4ebfd35bf643d27dbca1bb228957008617) (https://android.googlesource.com/toolchain/llvm 3c393fe7a7e13b0fba4ac75a01aa683d7a5b11cd) (based on LLVM 8.0.7svn)
  Target: armv7-none-linux-android16
  Thread model: posix
  InstalledDir: /opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin
  clang++: note: diagnostic msg: PLEASE submit a bug report to https://github.com/android-ndk/ndk/issues and include the crash backtrace, preprocessed source, and associated run script.
  clang++: note: diagnostic msg: 
  ********************
  
  PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
  Preprocessed source(s) and associated run script(s) are located at:
  clang++: note: diagnostic msg: /tmp/hlslParseHelper-3e494b.cpp
  clang++: note: diagnostic msg: /tmp/hlslParseHelper-3e494b.sh
  clang++: note: diagnostic msg: 
  
  ********************
  ninja: build stopped: subcommand failed.
  


* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 6.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/5.4.1-20190522110022+0000/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 14s
15 actionable tasks: 1 executed, 14 up-to-date
```

**Comment by unknownbrackets** [comment]
That's odd.  Seems like a clang bug?  Maybe check disk/storage space on the builder?

-[Unknown]

**Comment by devopsdeluxe** [comment]
I did a bit of digging and it looked quite similar to: https://github.com/android-ndk/ndk/issues/732, so I followed the crash's instructions:

```
  clang++: note: diagnostic msg: PLEASE submit a bug report to https://github.com/android-ndk/ndk/issues and include the crash backtrace, preprocessed source, and associated run script.
  clang++: note: diagnostic msg: 
  ********************
  
  PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
  Preprocessed source(s) and associated run script(s) are located at:
  clang++: note: diagnostic msg: /tmp/hlslParseHelper-3e494b.cpp
  clang++: note: diagnostic msg: /tmp/hlslParseHelper-3e494b.sh
  clang++: note: diagnostic msg: 
  
  ********************
```

and opened a new issue: https://github.com/android-ndk/ndk/issues/1000

**Comment by grravity** [comment]
Can we please update the F-Droid version

**Comment by devopsdeluxe** [comment]
I've updated the merge request: https://gitlab.com/fdroid/fdroiddata/-/merge_requests/3083

Awaiting final review.

**Comment by devopsdeluxe** [comment]
Updated my MR on GitLab to v1.10.2: https://gitlab.com/fdroid/fdroiddata/-/merge_requests/3083

**Comment by 0x647262** [comment]
Unfortunately, I'm going to be dropping this since F-Droid requires FFMPEG, com.bdacontroller.jar, and Vulkan to be built at compile time (something I currently don't have the knowledge & time to tackle). If someone knowledgeable on how to build those from source using PPSSPP's current build setup comes along, I would be happy to help them get the F-Droid version updated.

**Comment by MrHeating** [comment]
Thynk you anyway for putting your time in!
-- 
Sent with K-9 Mail. Please excuse my brevity.

**Comment by seniorm0ment** [comment]
Something @IzzyOnFDroid might have some knowledge on?

**Comment by seniorm0ment** [comment]
Something @IzzySoft might have some knowledge on?
Would be very nice to see it updated. The last FDroid release was 2015.

**Comment by IzzySoft** [comment]
I can just say that automated updates for the app at F-Droid are currently disabled as there were build issues (according to the notes: "android-arm64-build.sh: build_arm64: not found"). I'm not an Android dev so I cannot tell exactly – but you could open an issue at fdroiddata asking to have it updated or, even better, a merge request for `metadata/org.ppsspp.ppsspp.yml` having the latest release added in a way that it builds.

**Comment by 0x647262** [comment]
@seniorm0ment I currently don't have time to take a second swing at this, but I would be happy to point you in the right direction to get (almost) everything working. If you're interested, I can rebase my old MR on F-Droid's Gitlab and let you take a whack at it from there?

**Comment by barfin** [comment]
Any news on this?

**Comment by jugendhacker** [comment]
Hey, another F-Droid contributor here. Tried to get this building again, but somehow your fork of ffmpeg seems to have difficulties building with the prebuilt clang of the Android NDK. Maybe someone could help understanding [the problem](https://gitlab.com/jugendhacker/fdroiddata/-/jobs/3543992702#L616)?

**Comment by unknownbrackets** [comment]
@jugendhacker add this to your script:
```
export NDK_PLATFORM=$NDK/platforms/android-16/arch-arm
```

For arm64, use:
```
export NDK_PLATFORM=$NDK/platforms/android-21/arch-arm64
```

If you need x86/x86_64, it's the same as the above with "arm" replaced with "x86" and "arm64" replaced with "x86_64" (notably, 16 and 21, like above.)  Newer platforms are I guess okay, but might change minimum requirements.

That said, it also looks like we should remove the `--sysroot` args for the latest NDK.  I'll make a quick PR.

-[Unknown]

**Comment by unknownbrackets** [comment]
Once #16705 is merged, you should be able to use those exports without any other changes.

-[Unknown]

**Comment by jugendhacker** [comment]
@unknownbrackets thanks for your hint with the exports, this definitely helped a bit. But now I'm running into the issue described [here](https://github.com/hrydgard/ppsspp-ffmpeg/issues/59#issuecomment-793341282), looks like somehow the build process is missing some include paths...

**Comment by unknownbrackets** [comment]
Was that with my changed build scripts?  I got exactly the same error until I removed `--sysroot`, which should already be merged now.

-[Unknown]

**Comment by jugendhacker** [comment]
@unknownbrackets okay so this helped the build a bit further again, but now the linker seems to have problems: https://gitlab.com/jugendhacker/fdroiddata/-/jobs/3544478878#L6573

Maybe because it was built against a different platform? Also somehow interesting that the arm64 gradle build just worked fine...

**Comment by unknownbrackets** [comment]
Ah, you might have to increase the minSdkVersion.  I'm not really sure if there's a way to build NDK now with android-9 support...

-[Unknown]

**Comment by jugendhacker** [comment]
@unknownbrackets I tried patching minSdkVersion inside build.gradle and also bumping NDK_PLATFORM to android-21 for the armv7 builds too. This did not help naything. No I'm wondering how the binaries over hrydgard/ppsspp-ffmpeg got build exactly? Maybe I'm using the wrong ndk version or something.

**Comment by unknownbrackets** [comment]
I'm not sure, they were last updated in 2016 using ndk r12b or older.  It might be problematic to build for the latest NDK...

https://stackoverflow.com/questions/51767214/android-studio-with-ndk-link-error-undefined-reference-to-stderr
https://stackoverflow.com/questions/39322852/android-ndk-undefined-reference-to-stderr
https://github.com/android/ndk/issues/445#issuecomment-313322546

Maybe it's this last one.  We might need to add `-D__ANDROID_API__=21` or similar to the ffmpeg build scripts.

-[Unknown]

**Comment by jugendhacker** [comment]
This is quite unfortunate, especially if noone really knows how to build them anymore. I would recommend getting this working again, because without it F-Droid sadly can't update PPSSPP

**Comment by unknownbrackets** [comment]
Sorry, as explained here: https://github.com/hrydgard/ppsspp/issues/15308#issuecomment-1030655799

We infrequently have had a need to recompile it.  That said, as noted in #15969, it'd be nice to compile it each time on all platforms.  Currently we don't, especially because it's often a huge pain to get working for Android and Windows.

Did you try `-D__ANDROID_API__=21` on the Android FFmpeg builds?  It would go under `--extra-clags` as in:
```
    --extra-cflags=" --target=aarch64-none-linux-android21 -no-canonical-prefixes -fdata-sections -ffunction-sections -fno-limit-debug-info -funwind-tables -fPIC -O2 -DANDROID -DANDROID_PLATFORM=$BUILD_ANDROID_PLATFORM -Dipv6mr_interface=ipv6mr_ifindex -fasm -fno-short-enums -fno-strict-aliasing -Wno-missing-prototypes -D__ANDROID_API__=21" \
```

We actually wouldn't want to do these things, because we want to keep PPSSPP running on older Android devices, like the Xperia Play... I hope there's some way to get #15969 without jettisoning older devices.

-[Unknown]

**Comment by linsui** [comment]
> Did you try -D__ANDROID_API__=21 on the Android FFmpeg builds?

This works. :+1:

**Comment by linsui** [comment]
Updated.

**Comment by Florin9doi** [comment]
The metadata could get some updates, but this issue is mostly complete and can be closed.:
```
yml:
Translation: https://www.ppsspp.org/docs/development/translation/
Changelog: https://www.ppsspp.org/news/

files:
icon.png : https://github.com/hrydgard/ppsspp/blob/master/android/res/drawable-xxhdpi/ic_launcher.png
featureGraphic.png : https://github.com/hrydgard/ppsspp/blob/master/android/res/drawable-xhdpi/ic_banner.png
phoneScreenshots/*.png : ...
```

**Comment by linsui** [comment]
@Florin9doi You can add those files in fastlane structure in this repo. It will be fetched automatically in the next version.

**Comment by Florin9doi** [comment]
The Fdroid version don't have support for "Moga" bluetooth controllers

**Comment by Nitrovenom** [comment]
> Does the F-Droid version differ in any way from the Github APK (other than signing keys used)?

F-Droid one is larger in size

**Comment by hrydgard** [comment]
Should be no difference really. Not sure how they build the APK though.

**Comment by linsui** [comment]
It's 36.2MB vs 35.5MB. F-Droid builds ffmpeg with ndk r21e so it's normal the binary is a little larger.

**Comment by Nitrovenom** [comment]
Does it effect performance though?

**Comment by hrydgard** [comment]
Ah ok. 

I'm going to switch everything to a newer NDK in  the near future too, as part of #19658 .

**Comment by Nitrovenom** [comment]
> Does the F-Droid version differ in any way from the Github APK (other than signing keys used)?

![Image](https://github.com/user-attachments/assets/c5080584-a0f9-4713-92b2-a4f00742ba0c)

**Comment by hrydgard** [comment]
@Nitrovenom  It shouldn't differ.

**Comment by Nitrovenom** [comment]
1.19.x not released in fd. Hope you guys are gonna take action regarding this.

**Comment by hrydgard** [comment]
What kind of action exactly? I am not currently doing the f-droid releases, and I don't really know who is.

**Comment by Nitrovenom** [comment]
Maybe its is @linsui

**Comment by linsui** [comment]
Yes, but I don't use ppsspp so I don't check this regularly. If we can enable autoupdate then it will be updated in time. @hrydgard Currently the only thing preventing us from enabling autoupdate is that we need to get the version code from somewhere. We can only extract it with regex so we can't calculate it from the tag.

**Comment by linsui** [comment]
Or @Nitrovenom since you use it can you maintain the F-Droid metadata?

**Comment by Nitrovenom** [comment]
> Or [@Nitrovenom](https://github.com/Nitrovenom) since you use it can you maintain the F-Droid metadata?

I don't know how to.

**Comment by hrydgard** [comment]
@linsui do you mean that you need a checked-in file with the version, like say a version.txt containing "v1.19.3" in the current case? Or what exactly do you need? 

I can add such a file if it's needed, but it's kind of a pain too - would be better if you could just grab the tag, I don't understand why that wouldn't be possible.

**Comment by linsui** [comment]
We can extract the version name from the tag. But we can't extract the version code, i.e., 119030000. A version.txt works. It can be added to the repo, or github release. Or just put the number in the release note besides "PPSSPP is a cross-platform PSP emulator....".

**Comment by hrydgard** [comment]
The version code is generated from the tag. First digit is the major version, next two is the minor version (zero-padded to two digits), next is the "sub-minor" version (zero-padded to two digits). The last four digits are the commit number, which on tags will always be zero. 

So whatever script you're using can just do that, and it'll match.

**Comment by linsui** [comment]
Yes, I know I can calculate it. But F-Droid can only match it with regex. It doesn't support custom script.

**Comment by NABN00B** [comment]
> If [UpdateCheckMode](https://f-droid.org/en/docs/Build_Metadata_Reference/#UpdateCheckMode) is set to Tags, this [AutoUpdateMode] should be set to Version without any pattern. The checked tag is used directly.

https://f-droid.org/en/docs/Build_Metadata_Reference/#AutoUpdateMode



> Optionally append a regex pattern at the end - separated with a space - to only check the tags matching said pattern. Useful when apps tag non-release versions such as X.X-alpha, so you can filter them out ...

https://f-droid.org/en/docs/Build_Metadata_Reference/#UpdateCheckMode

We don't need the regex.

**Comment by linsui** [comment]
You misunderstand the doc or this issue. We need to extract the version code with regex so it has to be a static value in the file. [UpdateCheckMode](https://f-droid.org/en/docs/Build_Metadata_Reference/#UpdateCheckMode) is about which tag should be checked. [AutoUpdateMode](https://f-droid.org/en/docs/Build_Metadata_Reference/#AutoUpdateMode) is about how the new build block should be generated. The related field is UpdateCheckData.

**Comment by hrydgard** [comment]
I was hoping for an ugly hack from chatgpt, i've had luck with asking it for strange regexes before, but:

> is it possible to use a regex to convert a string like "v1.19.3" to 119030000 ?
> Note, the output format is ABBCC0000 where A is the major version, B is zero padded minor version, C is zero padded sub/patch version, and the zeroes are just always there.


> ChatGPT said:
> Yes, it's possible to use a regex to extract the parts, and then use code (like in Python, JavaScript, etc.) to reformat them into your desired format: ABBCC0000.

> However, regex alone can’t do zero-padding or arithmetic (like converting "1" to "01"), but it can extract the major, minor, and patch versions, and you can then format them as needed in code.


I don't understand why you are limited to a regex though - or even why you'd need the version in the first place. It can be parsed out of the APK.

**Comment by linsui** [comment]
F-Droid uses the version code to check if there is a new version, then update the metadata accordingly. The tool can only do regex match to get the version code.

**Comment by NABN00B** [comment]
```
AutoUpdateMode: Version
UpdateCheckMode: Tags
UpdateCheckData: '|(\d)(\d{2})(\d{2})(\d{4})||v(\d+)\.(\d+)(?:\.(\d+))?(?:-(\d+))?'
```

Does this not suffice? 🤔

**Comment by linsui** [comment]
It can only have one match group.

**Comment by NABN00B** [comment]
```
AutoUpdateMode: Version
UpdateCheckMode: Tags [^-]*
UpdateCheckData: '|(\d+)||v([\d.]+)'
```

This? 😄

**Comment by NABN00B** [comment]
The [documentation for UpdateCheckData](https://f-droid.org/en/docs/Build_Metadata_Reference/#UpdateCheckData) explicitly mentions that

```
Examples for UpdateCheckMode: Tags:

...

- If no file for the version code was specified, code and name can be extracted from the tag: '|\+(\d+)||Android-([\d.]+)'

```

but I have no clue where the `+` and the `Android-` come from.

EDIT: now I understand it, this example is for a repo whose tags are `Android-versionName+versionCode`.

**Comment by linsui** [comment]
No, as the AI said, regex alone can’t do zero-padding or arithmetic.

**Comment by Nitrovenom** [comment]
> No, as the AI said, regex alone can’t do zero-padding or arithmetic.

Update fdroid version to 1.20.3 plz

**Comment by linsui** [comment]
@hrydgard Can you also update the script for armeabi-v7a and x86_64 for NDK 29 as in https://github.com/hrydgard/ppsspp-ffmpeg/commit/2e90f7ac30425659a60a1497a87353d696b285c0?

**Comment by asl97** [comment]
> Update fdroid version to 1.20.3 plz

Seem like there was an attempt: https://gitlab.com/fdroid/fdroiddata/-/merge_requests/35392

But (I assume) the F-droid version is stalled due to some compile issue?

**Comment by hrydgard** [comment]
No clue what's going on there...

**Comment by linsui** [comment]
You only updated the script for arm64. We also need to build other abis.

**Comment by hrydgard** [comment]
Me? I didn't touch it.

**Comment by linsui** [comment]
See the comment above. https://github.com/hrydgard/ppsspp-ffmpeg/commit/2e90f7ac30425659a60a1497a87353d696b285c0 This makes the arm64 lib build.

**Comment by hrydgard** [comment]
Oh, sorry, I missed that. Will do, although I'm not sure why that would help - it currently builds fine on modern SDK linking to the old libraries, for me at least...
